### PR TITLE
Achievement updates

### PR DESCRIPTION
**Pull request by sderickson**
Have the users periodically check for new achievements and check if they've been earned, so that they automatically get awarded new achievements retroactively.

Also refactored the POST /db/earned_achievement endpoint and some related logic.

For now, the client DOES NOT actually check for achievements to award retroactively. I'd like to deploy the endpoint and test it through a dev proxy to ensure it works correctly.

### CODE REVIEW

File: server/models/EarnedAchievement.coffee [hunk 1]
```diff
@@ -16,81 +17,106 @@ EarnedAchievementSchema.pre 'save', (next) ->
...
+    return earned
+    
+    
+EarnedAchievementSchema.statics.createForAchievement = co.wrap (achievement, doc, options={}) ->
+  { previouslyEarnedAchievement, originalDocObj } = options
+  
+  User = require('./User')
   userObjectID = doc.get(achievement.get('userField'))
-  userID = if _.isObject userObjectID then userObjectID.toHexString() else userObjectID # Standardize! Use strings, not ObjectId's
+  userID = if _.isObject userObjectID then userObjectID.toHexString() else userObjectID # Standardize! Use ObjectIds
```
Comment (nwinter) [inline server/models/EarnedAchievement.coffee]: Now the comment says to use ObjectIds but we still are changing them to strings?

File: app/core/application.coffee [hunk 2]
```diff
@@ -44,42 +44,53 @@ window.console ?=
...
+    @tracker = new Tracker()
+    @facebookHandler = new FacebookHandler()
+    @gplusHandler = new GPlusHandler()
+    @githubHandler = new GitHubHandler()
+    @moduleLoader = new ModuleLoader()
+    @moduleLoader.loadLanguage(me.get('preferredLanguage', true))
+    $(document).bind 'keydown', preventBackspace
+    preload(COMMON_FILES)
+    CocoModel.pollAchievements()
+#    @checkForNewAchievement()
```
Comment (nwinter) [inline app/core/application.coffee]: Supposed to be in or deleted?

File: app/core/application.coffee [hunk 3]
```diff
@@ -44,42 +44,53 @@ window.console ?=
...
-    onIdleChanged = (to) => => Backbone.Mediator.publish 'application:idle-changed', idle: @userIsIdle = to
-    @idleTracker = new Idle
-      onAway: onIdleChanged true
-      onAwayBack: onIdleChanged false
-      onHidden: onIdleChanged true
-      onVisible: onIdleChanged false
-      awayTimeout: 5 * 60 * 1000
-    @idleTracker.start()
+Application = {
+  initialize: ->
```
Comment (sderickson) [inline app/core/application.coffee]: Make into a class

File: server/models/EarnedAchievement.coffee (hunk 1 above)
Comment (sderickson) [inline server/models/EarnedAchievement.coffee]: Yeah, until we migrate everything that's already in the db. I'll clarify the comment, it should be a TODO.

File: app/core/application.coffee (hunk 2 above)
Comment (sderickson) [inline app/core/application.coffee]: Commented out for now. See the PR description; I'm going to deploy it without triggering it generally until I've tested through proxy.

File: server/models/EarnedAchievement.coffee [hunk 4]
```diff
@@ -16,81 +17,106 @@ EarnedAchievementSchema.pre 'save', (next) ->
...
+  if achievement.get('proportionalTo') and earned
+    earnedAchievementDoc = yield @createForAchievement(achievement, trigger, {previouslyEarnedAchievement: earned})
+    return earnedAchievementDoc or earned
+
+  else if earned
+    achievementEarned = achievement.get('rewards')
+    actuallyEarned = earned.get('earnedRewards')
+    if not _.isEqual(achievementEarned, actuallyEarned)
+      earned.set('earnedRewards', achievementEarned)
+      yield earned.save()
```
Comment (sderickson) [inline server/models/EarnedAchievement.coffee]: This bit should also update the user

**Pull request merged**